flink-connectors/flink-connector-base/src/main/java/org/apache/flink/connector/base/source/hybrid/HybridSourceReader.java

```diff
@@ -0,0 +1,251 @@
+    private List<HybridSourceSplit> restoredSplits = new ArrayList<>();
+
+    public HybridSourceReader(
+            SourceReaderContext readerContext, Map<Integer, Source> switchedSources) {
+        this.readerContext = readerContext;
+        this.switchedSources = switchedSources;
+    }
+
+    @Override
+    public void start() {
+        // underlying reader starts on demand with split assignment
+        int initialSourceIndex = currentSourceIndex;
+        if (!restoredSplits.isEmpty()) {
+            initialSourceIndex = restoredSplits.get(0).sourceIndex() - 1;
+        }
+        readerContext.sendSourceEventToCoordinator(
+                new SourceReaderFinishedEvent(initialSourceIndex));
+    }
+
+    @Override
+    public InputStatus pollNext(ReaderOutput output) throws Exception {
+        if (currentReader == null) {
+            return InputStatus.NOTHING_AVAILABLE;
+        }
+
+        InputStatus status = currentReader.pollNext(output);
+        if (status == InputStatus.END_OF_INPUT) {
+            // trap END_OF_INPUT unless all sources have finished
+            LOG.info(
+                    "End of input subtask={} sourceIndex={} {}",
+                    readerContext.getIndexOfSubtask(),
+                    currentSourceIndex,
+                    currentReader);
+            // Signal the coordinator that this reader has consumed all input and the
+            // next source can potentially be activated.
+            readerContext.sendSourceEventToCoordinator(
+                    new SourceReaderFinishedEvent(currentSourceIndex));
+            if (!isFinalSource) {
+                // More splits may arrive for a subsequent reader.
+                // InputStatus.NOTHING_AVAILABLE suspends poll, requires completion of the
+                // availability future after receiving more splits to resume.
+                if (availabilityFuture != null && availabilityFuture.isDone()) {
+                    // reset to avoid continued polling
+                    availabilityFuture = new CompletableFuture();
+                }
+                return InputStatus.NOTHING_AVAILABLE;
+            }
+        }
+        return status;
+    }
+
+    @Override
+    public List<HybridSourceSplit> snapshotState(long checkpointId) {
+        List<? extends SourceSplit> state =
+                currentReader != null
+                        ? currentReader.snapshotState(checkpointId)
+                        : Collections.emptyList();
+        return HybridSourceSplit.wrapSplits(currentSourceIndex, state);
+    }
+
+    @Override
+    public CompletableFuture<Void> isAvailable() {
+        // track future to resume reader after source switch
+        if (currentReader != null) {
+            return availabilityFuture = currentReader.isAvailable();
+        } else {
+            LOG.debug("Suspending pollNext due to no underlying reader");
+            return availabilityFuture = new CompletableFuture<>();
```

Comment:
I'd not create a new future here but when having reached `currentReader = null` at the switching code.
If there are two places that use is `isAvailable` you may end up with 2 futures of which you only ever complete one.

Reply:
Done, it is now created during initialization.